src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using System.Web.Hosting;
+using System.Web.Http;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Timers;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public class WebScriptHostExceptionHandler : WebJobExceptionHandler
+    {
+        /// <summary>
+        /// If the exception is a <see cref="TimeoutException"/>, log details and gracefully shut down.
+        /// Otherwise, rethrow the exception from another thread to kill the process.
+        /// </summary>
+        /// <param name="exceptionInfo">The <see cref="ExceptionDispatchInfo"/> to handle.</param>
+        public override void Handle(ExceptionDispatchInfo exceptionInfo)
+        {
+            if (IsTimeout(exceptionInfo))
+            {
+                WebScriptHostManager webScriptHostManager = GlobalConfiguration.Configuration.DependencyResolver.GetService<WebScriptHostManager>();
+                webScriptHostManager?.Instance.TraceWriter.Error("A function timeout has occurred. Host is shutting down.");
+
+                // Give the manager some time to shut down gracefully.
+                Task shutdownTask = Task.Delay(2000).ContinueWith((ct) =>
+                {
+                    HostingEnvironment.InitiateShutdown();
+                });
+
+                webScriptHostManager?.Stop();
+
+                // we don't want to exit before we know shutdown has been initiated
+                shutdownTask.Wait();
+                return;
+            }
+
+            base.Handle(exceptionInfo);
+        }
+
+        private static bool IsTimeout(ExceptionDispatchInfo exceptionInfo)
```

Comment:
Since I won't be making use of either of these helpers with this commit, I didn't do that with these changes.

---
In reply to: [76696680](https://github.com/Azure/azure-webjobs-sdk-script/pull/604#discussion_r76696680) [](ancestors = 76696680)